modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -974,6 +997,42 @@ private CompletableFuture<List<BinaryRow>> retrieveExactEntriesUntilCursorEmpty(
         });
     }
 
+    private static void mergeRowsWithResolvedWriteIntents(
+            List<BinaryRow> rows,
+            Int2ObjectMap<CompletableFuture<TimedBinaryRow>> resolutionFutures,
+            List<BinaryRow> result
+    ) {
+        int curIndex = 0;
+        int rowsIndex = 0;
+
+        for (Entry<CompletableFuture<TimedBinaryRow>> entry : resolutionFutures.int2ObjectEntrySet()) {
```

Comment:
What if we pre-inserted `null` into an original list, and then instead of joining sub-lists we would just do a `set` in a loop? Such an algorithm seems much more simple to me.
Why have you decided to make it this complicated?

Another option, even more efficient one. Have a `List<Object>`. When we're done, replace each completable future value with its result, and cast this list to `List<BinaryRow>`. A little bit dirty, but with a good comment it might be the best solution

Reply:
> What if we pre-inserted null into an original list, and then instead of joining sub-lists we would just do a set in a loop? 

This seems not possible, because a write intent can resolve into a `null` and we do not return nulls